src/Common/src/TypeSystem/TypesDebugInfoWriter/UserDefinedTypeDescriptor.cs

```diff
@@ -0,0 +1,162 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Reflection.Metadata;
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.TypeSystem.TypesDebugInfo
+{
+    public class UserDefinedTypeDescriptor
+    {
+        public UserDefinedTypeDescriptor(ITypesDebugInfoWriter objectWriter)
+        {
+            _objectWriter = objectWriter;
+        }
+        public uint GetTypeIndex(TypeDesc type)
+        {
+            uint variableTypeIndex = 0;
+            if (!_knownTypes.TryGetValue(type, out variableTypeIndex))
+            {
+                variableTypeIndex = GetNewTypeIndex(type);
+            }
+            return variableTypeIndex;          
+        }
+
+        private uint GetNewTypeIndex(TypeDesc type)
+        {
+            switch (type.Category)
```

Comment:
Interfaces and polymorph classes don't work well now, it will be implemented in the future.
However, such test showed a problem with step-in into method through virtual or interface table. I will take a look on  and create an issue for that.